apps/ledger-live-mobile/src/screens/SyncOnboarding/EarlySecurityCheck.tsx

```diff
@@ -1,42 +1,405 @@
+        // But an error `CantOpenDevice` should be triggered quickly after.
+        log("EarlySecurityCheck", "Genuine check permission state set to unlock-needed");
+      } else if (devicePermissionState === "requested") {
+        currentDisplayedDrawer = "allow-manager";
+      } else if (devicePermissionState === "refused") {
+        currentDisplayedDrawer = "none";
+        genuineCheckUiStepStatus = "genuineCheckRefused";
+      }
+    } else if (genuineCheckStatus === "error") {
+      // Currently genuine check error or refused is handled in the same way. This can be changed in the future.
+      currentDisplayedDrawer = "genuine-check-error";
+    } else if (genuineCheckStatus === "non-genuine") {
+      currentDisplayedDrawer = "non-genuine";
+    }
+  }
+  // `currentStep` can be any value for those UI updates
+  if (genuineCheckStatus === "completed") {
+    genuineCheckUiStepStatus = "completed";
+  } else if (["error", "non-genuine"].includes(genuineCheckStatus)) {
+    genuineCheckUiStepStatus = "error";
+  }
+
+  // Handles firmware update check UI logic
+  if (currentStep === "firmware-update-check") {
+    if (firmwareUpdateCheckStatus === "ongoing") {
+      firmwareUpdateUiStepStatus = "active";
+      currentDisplayedDrawer = "none";
+
+      // Updates the firmwareUpdateCheckStatus
+      // If the current error triggered a retry attempt, does not display failure
+      if (
+        getLatestAvailableFirmwareError &&
+        !(
+          getLatestAvailableFirmwareError.type === "SharedError" &&
+          getLatestAvailableFirmwareError.retrying
+        )
+      ) {
+        log(
+          "EarlySecurityCheck",
+          "Failed to retrieve latest firmware version with error:",
+          getLatestAvailableFirmwareError.name,
+        );
+        setFirmwareUpdateCheckStatus("error");
+      } else if (getLatestAvailableFirmwareStatus === "no-available-firmware") {
+        setFirmwareUpdateCheckStatus("completed");
+      }
+
+      if (getLatestAvailableFirmwareLockedDevice) {
+        // The device has no PIN in the ESC.
+        // Plus if the error is an `UnresponsiveDeviceError`: it would probably means the user
+        // cancelled a firmware update previously and came back to the ESC (with a device in an incorrect state).
+        log(
+          "EarlySecurityCheck",
+          "Device considered locked while getting latest available firmware",
+          { error: getLatestAvailableFirmwareError },
+        );
+      }
+
+      // Updates the UI
+      if (getLatestAvailableFirmwareStatus === "available-firmware" && latestFirmware) {
+        // Only for QA to have always the same 1-way path: on infinite loop firmware
+        // the path 1.2.1-il2 -> 1.2.1-il0 does not trigger a firmware update during the ESC
+        if (
+          getLatestAvailableFirmwareStatus === "available-firmware" &&
+          latestFirmware?.final.name === "1.2.1-il0" &&
+          deviceInfo?.version === "1.2.1-il2"
```

Comment:
maybe make it work like that for all possible versions of IL fw?

Reply:
perfectionniste 🫣 done ;)